### PR TITLE
[#276]Support WAL shipping to ssh storage engine

### PR DESCRIPTION
**Pull request by HazemRawi**

### CODE REVIEW

**Comment by Hazem-404** [comment]
> 

Very Good Code Hazem !

File: src/include/storage.h [hunk 1]
```diff
@@ -63,6 +69,26 @@ pgmoneta_storage_create_s3(void);
...
 pgmoneta_storage_create_azure(void);
 
+/**
+ * Open WAL shipping file in remote ssh server
+ * @param srv The server index
+ * @param filename WAL file name
+ * @param segsize WAL segment size
+ * @return sftp_file WAL streaming file
+ */
+sftp_file
```
Comment (jesperpedersen) [inline src/include/storage.h:79]: Use an out parameter instead, and `int` as the return type with the status of the operation

File: src/include/utils.h [hunk 2]
```diff
@@ -861,6 +861,16 @@ pgmoneta_permission_recursive(char* d);
...
 pgmoneta_permission(char* e, int user, int group, int all);
 
+/**
+ * Calculate mode
+ * @param user The user (0; nothing, 4; read, 6; read/write)
+ * @param group The group (0; nothing, 4; read, 6; read/write)
+ * @param all All (0; nothing, 4; read, 6; read/write)
+ * @return The mode
+ */
+mode_t
```
Comment (jesperpedersen) [inline src/include/utils.h:871]: Same here

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
There are some formatting changes - use `Uncrustify_d-0.77.1_f`

**Comment by jesperpedersen** [comment]
Generally, I think this is going in the right direction... but I need to take a closer look

**Comment by GuChad369** [comment]
When GitHub runs the environment for macOS on this PR, it uses `/opt` as the global directory, whereas normally it should be `/usr`, which would already have the `ssh` library installed. I'm not sure why this change occurred. However, you can modify the `.github/workflows/ci.yml` file to install the library, or manually locate the library to ensure it works on macOS.

**Comment by HazemRawi** [comment]
@GuChad369 Thanks for your help! 
Jesper handled this problem

**Comment by HazemRawi** [comment]
I tested both workflows:

   -When the WAL segment file already exists.
   -When the WAL segment file needs to be created.

Everything works fine except that `sftp_wal_prepare()` takes about 12 minutes to execute. I suspect this delay might be due to the server(E2 server from Google cloud).
Do you have any ideas on how to improve this?

**Comment by jesperpedersen** [comment]
Did you test it with a SSH server ? Or on your network ?

**Comment by HazemRawi** [comment]
> Did you test it with a SSH server ? Or on your network ?

I test it using SSH server as described on `ssh.md`

>I suspect this delay might be due to the server

I mean the server doesn't come with a high system specification, As most of the time taken by `sftp_write() `.

**Comment by jesperpedersen** [comment]
Ok, understood

**Comment by jesperpedersen** [comment]
Rebased, and merged.

Thanks for your contribution !

**Pull request closed without merging**